Step1 - 엔티티 매핑 PR 요청드립니다.

KangJunJun opened the pull request.

KangJunJun committed:
chore: 쿼리 확인을 위한 Spring, JPA 프로퍼티 추가

KangJunJun committed:
docs: README.md 요구사항 및 맵핑할 엔티티 리스트업

KangJunJun committed:
feat: 각 도메인 Entity 설정

- 최초 기본동작 및 테스트 가능하도록 각 도메인에 처리함

KangJunJun committed:
feat: BaseEntity 생성

- 공통으로 사용되는 필드 속성을 상속 받도록 추가

KangJunJun committed:
chore: dialect 설정 제거

- 불필요한 dialect 설정으로 인해 SQLGrammarException 에러 발생

KangJunJun committed:
feat: Answer Entity 맵핑 및 테스트코드 추가

KangJunJun committed:
feat: DeleteHistory Entity 맵핑 추가

KangJunJun committed:
feat: Question Entity 맵핑 추가

KangJunJun committed:
fix: createdAt 필드에 not null 적용

KangJunJun committed:
fix: Question title 필드에 길이제한 추가

KangJunJun committed:
feat: User Entity 맵핑 추가

chae-yh commented on README.md:
```diff
@@ -0,0 +1,64 @@
+# JPA
+## STEP 1 - 엔티티 맵핑
```
단순히 요구사항을 그대로 넣어주기보다는, 기능 요구 사항에 대해서 작성을 추가로 해보는 것은 어떨까요...?

chae-yh commented on src/main/java/qna/Application.java:
```diff
@@ -2,7 +2,8 @@
 import org.springframework.boot.autoconfigure.SpringBootApplication;
-
+import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
+@EnableJpaAuditing
```
Auditing 추가 👍

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -3,13 +3,16 @@
-public class Answer {
+@Entity
+public class Answer extends BaseEntity {
     private Long id;
```
id 값임을 명시해주고, id 생성 전략도 추가되면 좋을 것 같아요..!

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -3,13 +3,16 @@
-
-public class Answer {
+@Entity
+public class Answer extends BaseEntity {
```
JPA 사용시 기본생성자가 필요할 것 같습니다.
추가로, 작업해주신 클래스에 setter / getter가 그대로 있는데, 해당 메서드들은 언제 사용하게 될까요..?

KangJunJun replied:
setter / getter 를 언제 사용하게 될까요? 라고 물어봐주셨는데,
단위테스트를 하라는 의도인지 안쓰는건 삭제하라는 의도인지,  즉 어떤 의미인지 이해를 못했습니다 😢

chae-yh replied:
해당 내용 기록을 위해 남겨둡니다 ㅎㅎ

`우선 의도는 삭제였습니다 ㅎ.ㅎ;; getter 와 setter는 단순하게 데이터를 가져오고 set해주는 부분이라 단위테스트의 의미도 없을 것 같아요. 게다가 저희는 객체지향 생활체조 원칙에 맞게 개발을 진행하고 있고, 해당 원칙의 9번이 getter / setter를 쓰지 않는 부분입니다. 이러한 단순 메서드는 삭제하면 좋을 것 같아요,,,!`

chae-yh commented on src/main/java/qna/domain/BaseEntity.java:
```diff
@@ -0,0 +1,24 @@
+public abstract class BaseEntity {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
```
만약 특정 테이블에 id가 필요없거나 넣으면 안될 경우에는 해당 BaseEntity를 어떻게 사용해야 할까요..?

KangJunJun replied:
BaseEntity를 상속받는 엔티티는 모두 Id를 필요로 하기에 이렇게 추가했는데 엔티티에 명시를 하는게 좋을까요?

chae-yh replied:
기록을 위해 남겨둡니다...!

```
BaseEntity의 경우는 한번 만들어두면 뒤의 객체들이 추가될 때 자연스럽게 추가가 되고, 뒤에 수정하기가 매우 어렵
습니다. 생성시각 / 수정시각 같은 경우는 모든 엔티티에 추가되어도 크게 이슈가 없겠지만, id 같은 경우는 특정 엔티티의 경우에는 오히려 넣으면 안되는 경우가 있을 수 있을 것 같아요. 현재 구조에서는 BaseEntity에 id를 넣어도 괜찮겠지만 추후의 추가되는 모든 클래스에 대해서도 올바른 구조가 될까요...? 해당 이유에서 id는 각각의 엔티티에서 명시해주는 게 좋다고 생각합니다.
```

chae-yh commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,29 @@
     public static final Answer A2 = new Answer(UserTest.SANJIGI, QuestionTest.Q1, "Answers Contents2");
+    
+    @Test
+    void Answer_contents_test() {
```
```suggestion
    void answer_contents_test() {
```
메서드이므로 소문자로 작성해주면 좋을 것 같습니다. 추가로, `DisplayName으로 어떤 테스트인지 구체적으로 작성해주어도 좋을 것 같아요...!

chae-yh commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,29 @@
+    }
+
+    @Test
+    void Answer_isDeleted_test() {
```
테스트 중에서 `Answer_questionId_test()` 와 `Answer_isDeleted_test()` 의 의도가 어떤 부분인지 궁금합니다.
Answer_questionId_test() 에서는 null 값 비교이고, Answer_isDeleted_test()에서는 삭제 됐는지 여부를 두 개 객체간에 비교해주는데 어떤 의미가 있을까요...?

chae-yh commented on src/test/java/qna/domain/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,34 @@
+    
+    @Test
+    void save() {
+        Answer answerTest = AnswerTest.A1;
```
테스트 간 객체 사용이 분리되어야 할 것 같습니다. 
반대로 동일한 형태의 객체를 반복적으로 사용한다면[ Fixture로 분리해서 사용](https://jojoldu.tistory.com/611)하면 어떨까요...?

chae-yh commented on src/test/java/qna/domain/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,34 @@
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+@DataJpaTest
+public class AnswerRepositoryTest {
```
Repositor에 대한 슬라이스 테스트를 만들어주신 것 좋습니다 ㅎㅎ 👍 다만 테스트를 조금 더 다양하게 작성해볼 수 있을 것 같아요. 예를 들어, 저장한 Entity와 해당 id로 조회한 Entity의 동일성 등에 대한 테스트도 추가되면 좋을 것 같습니다.

chae-yh commented on src/test/java/qna/domain/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,34 @@
+    
+    @BeforeEach
+    void setUp() {
+        answerRepository.deleteAllInBatch();
```
혹시 해당 삭제를 해주는 이유 여쭤봐도 될까요...?

chae-yh commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,23 @@
+        assertThat(Q1.getTitle()).isEqualTo("title1");
+    }
+    @Test
+    void Question_contents_test() {
```
`Question_title_test()` 혹은 `Question_contents_test()`는 단순히 getter를 테스트하는 것인데 의미가 있는 테스트일까요...?

chae-yh commented on src/test/java/qna/domain/QuestionRepositoryTest.java:
```diff
@@ -0,0 +1,33 @@
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+@DataJpaTest
+public class QuestionRepositoryTest {
```
해당 테스트도 AnswerRepository와 비슷한 이슈가 있을 것 같습니다...!

chae-yh commented on src/main/java/qna/domain/User.java:
```diff
@@ -2,15 +2,19 @@
-
-public class User {
+@Entity
+public class User extends BaseEntity{
```
코드 템플릿이 적용되지 않은 것 같아요. [다른 교육생분들이 적용한 것](https://nextstep-members.slack.com/archives/C048F1DLU72/p1666930409544319)을 참고하여 적용부탁드립니다...!

KangJunJun replied:
코드템플릿이 뭔지 잘 모르겠습니다 가이드 부탁드립니다..!

chae-yh replied:
기록을 위해 남겨둡니다..!

[교육생 동기분이 정리해두신 부분](https://nextstep-members.slack.com/archives/C048F1DLU72/p1666930409544319)이 있는데, 해당 내용 확인 부탁드립니다...!

chae-yh reviewed (changes requested):
준호님 안녕하세요. JPA 미션을 같이 진행하게 될 채영훈이라고 합니다. 반갑습니다 ㅎㅎ
첫번째 미션, DDL문과 Entity간의 매핑 해주는 작업이었는데, 전체적으로 잘 해주신 것 같아요 👍 
하지만 조금 고민해주시면 좋을 것 같은 내용들이 있어 코멘트 드리니 확인 부탁드립니다.

감사합니다 : )

KangJunJun committed:
docs: 기능요구사항 변경

KangJunJun committed:
refactor: 각 엔티티에 id 컬럼 명시

KangJunJun committed:
refactor: 피드백 내용 적용

- 코드 템플릿 적용
- 의미 없는 단위테스트 제거
- 기본생성자 추가
- 사용안하는 getter setter 제거
- Test Fixture 사용

KangJunJun committed:
refactor: 저장한 엔티티의 id로 조회한 경우 동일한지 테스트하도록 수정

chae-yh commented on src/test/java/qna/domain/QuestionRepositoryTest.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Test
+    @DisplayName("Question 저장한 엔티티의 id로 조회한 경우 동일성 테스트")
+    void find() {
```
현재 상태로 동일성 테스트를 진행하면, 영속성 컨텍스트에 있는(캐시되어 있는) 동일한 객체를 가져와서 비교하기 때문에, Quenstion에 equals와 hashCode 재정의 한 부분이 없어도 정상 동작합니다.
실제로 해당 테스트 수행해보시면 insert 쿼리는 나오는데 select 쿼리가 나오지 않는 것으로 실제 db에 접근하지 않고 캐시된 값을 가져오는 것을 확인하실 수 있어요...!

혹시 시간이 되시면 영속성 컨텍스트를 비워서 실제 db에서 데이터를 가져와서 동등성 비교를 해보는 것도 좋을 것 같습니다. ㅎㅎ

chae-yh reviewed (approved):
준호님 안녕하세요. 피드백 반영 잘해주셨습니다 ㅎㅎ 👍 
소소하게 확인해보시면 좋을만한 내용이 있어 코멘트 드렸으니 살짝 확인만 해주시면 좋을 것 같습니다.

다음 step 진행 부탁드립니다. 감사합니다 : )
